samza-core/src/main/java/org/apache/samza/zk/ScheduleAfterDebounceTime.java

```diff
@@ -51,10 +50,23 @@
 
   public static final int DEBOUNCE_TIME_MS = 2000;
 
+  private final ScheduledTaskFailureCallback scheduledTaskFailureCallback;
+
   private final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
       new ThreadFactoryBuilder().setNameFormat("debounce-thread-%d").setDaemon(true).build());
   private final Map<String, ScheduledFuture> futureHandles = new HashMap<>();
 
+  // Ideally, this should be only used for testing. But ZkBarrierForVersionUpgrades uses it. This needs to be fixed.
+  // TODO: Timer shouldn't be passed around the components. It should be associated with the JC or the caller of
+  // coordinationUtils.
+  public ScheduleAfterDebounceTime() {
```

Comment:
> I think the debounceTimer should be associated with the groupId of the coordinationUtils, so the same thread is used for all the action related to the same processor.

`ScheduleAfterDebounceTime` is the actual event processing loop of the `ZkJobCoordinator`. I don't want to couple this with any other component that may use `CoodinationUtils`, which is used as a **library**.  It should be left up-to the caller on how it deals with the asynchronous callbacks when using a library. 

> Also I don't where this constructor is invoked, so may be it can be removed.

This constructor is used in `getCoordinationService` of `ZkCoordinationServiceFactory`. I didn't want to re-factor this code along with this PR. We can do it as a part of SAMZA-1128

> it looks like errorHandling listener should be passed from the point of invocation, so it could be different for different components.

It could be different for different components. But at this point, we are differentiating the different error scenarios. That's why I didn't add it to the `schedule` api and, instead added it to the constructor. Let me try adding to the schedule method itself.